module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,308 @@
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			return
+		}
+
+		serializeErr = bcw.Flush()
+	}()
+
+	cids, recvErr := s.storeBlobs(ctx, br)
+
+	<-done
+
+	if recvErr != nil {
+		return nil, recvErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	s.metrics.ExecutionDataAddStarted()
+
+	start := time.Now()
+	cids, err := s.addBlobs(ctx, sd)
+
+	if err != nil {
+		s.metrics.ExecutionDataAddFinished(time.Since(start), false, 0)
+
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	var blobTreeNodes int
+
+	for {
+		blobTreeNodes += len(cids)
+
+		if len(cids) == 1 {
+			s.metrics.ExecutionDataAddFinished(time.Since(start), true, blobTreeNodes)
+
+			return cids[0], nil
+		}
+
+		if cids, err = s.addBlobs(ctx, cids); err != nil {
+			s.metrics.ExecutionDataAddFinished(time.Since(start), false, blobTreeNodes)
+
+			return cid.Undef, fmt.Errorf("failed to add cid blobs: %w", err)
+		}
+	}
+}
+
+func (s *ExecutionDataService) retrieveBlobs(parent context.Context, bs *BlobSender, cids []cid.Cid) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids)
+	cachedBlobs := make(map[cid.Cid]network.Blob)
+
+	for _, c := range cids {
+		var err error
+		blob, ok := cachedBlobs[c]
+
+		if ok {
+			delete(cachedBlobs, c)
+		} else {
+			blob, err = s.findBlob(blobChan, c, cachedBlobs)
+
+			if err != nil {
+				return err
+			}
+
+			if blob == nil {
+				// the blob channel may be closed as a result of the context being canceled,
+				// in which case we should return the context error.
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+
+				return &BlobNotFoundError{c}
+			}
+		}
+
+		if err = bs.Send(blob); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *ExecutionDataService) findBlob(
+	blobChan <-chan network.Blob,
+	target cid.Cid,
+	cache map[cid.Cid]network.Blob,
+) (network.Blob, error) {
+	for blob := range blobChan {
+		// check blob size
+		blobSize := len(blob.RawData())
+		if blobSize > s.maxBlobSize {
+			return nil, &BlobSizeLimitExceededError{blob.Cid()}
+		}
+
+		if blob.Cid() == target {
+			return blob, nil
+		}
+
+		cache[blob.Cid()] = blob
+	}
+
+	return nil, nil
+}
+
+func (s *ExecutionDataService) getBlobs(ctx context.Context, cids []cid.Cid) (interface{}, error) {
+	bcr, bs := OutgoingBlobChannel()
+
+	done := make(chan struct{})
+	var v interface{}
+	var deserializeErr error
+
+	go func() {
+		defer close(done)
+		defer bcr.Close()
+
+		v, deserializeErr = s.serializer.Deserialize(bcr)
+	}()
+
+	sendErr := s.retrieveBlobs(ctx, bs, cids)
+
+	<-done
+
+	if sendErr != nil && errors.Is(sendErr, ErrClosedBlobChannel) {
+		return nil, sendErr
+	}
```

Comment:
what if `sendErr` is not a `ErrClosedBlobChannel` or nil?

Reply:
did you mean `!errors.Is(sendErr, ErrClosedBlobChannel)`?

Reply:
26c2d03

Reply:
yes, that's what I meant, good catch